ports-of-call/span.hpp

```diff
@@ -0,0 +1,464 @@
+      : ptr(p_ptr), size(p_size) {}
+
+  E *ptr = nullptr;
+  std::size_t size = 0;
+};
+
+// some metaprogramming
+template <class T>
+using uncvref_t = typename std::remove_cv<typename std::remove_reference<T>::type>::type;
+
+// detector of span
+template <class Q>
+struct is_span_oracle : std::false_type {};
+
+template <class T, std::size_t Extent>
+struct is_span_oracle<span<T, Extent>> : std::true_type {};
+
+template <class Q>
+struct is_span : is_span_oracle<typename std::remove_cv<Q>::type> {};
+
+template <class Q>
+struct is_std_array_oracle : std::false_type {};
+
+template <class T, std::size_t Extent>
+struct is_std_array_oracle<std::array<T, Extent>> : std::true_type {};
+
+template <class Q>
+struct is_std_array : is_std_array_oracle<typename std::remove_cv<Q>::type> {};
+
+template <class Q>
+struct is_array : std::false_type {};
+
+template <class T>
+struct is_array<T[]> : std::true_type {};
+
+template <class T, std::size_t N>
+struct is_array<T[N]> : std::true_type {};
+
+// detect C.size() and C.data()
+template <class, class = void>
+struct has_size_and_data : std::false_type {};
+
+template <class C>
+struct has_size_and_data<C, std::void_t<decltype(std::size(std::declval<C>())),
+                                        decltype(std::data(std::declval<C>()))>>
+    : std::true_type {};
+
+// detect if *(C.data()) is compatible with E
+template <class, class, class = void>
+struct is_compatible_element : std::false_type {};
+
+template <class C, class E>
+struct is_compatible_element<C, E, std::void_t<decltype(std::data(std::declval<C>()))>>
+    : std::is_convertible<typename std::remove_pointer<decltype(std::data(
+                              std::declval<C &>()))>::type (*)[],
+                          E (*)[]> {};
+
+// A container C:
+// - is not any of [span, c-array, std::array]
+// - has C.size() and C.data() interface
+template <class C>
+struct is_container
+    : std::bool_constant<!is_span<C>::value && !is_array<C>::value &&
+                         !is_std_array<C>::value && has_size_and_data<C>::value> {};
+
+// detect if C is a container and is compatible with element E
+template <class C, class E>
+struct is_compatible_container
+    : std::bool_constant<is_container<C>::value && is_compatible_element<C, E>::value> {};
+
+// detect if T is a complete type
+// NB: sizeof(T) cannot be used with incomplete types
+template <class, class = std::size_t>
+struct is_complete : std::false_type {};
+
+template <class T>
+struct is_complete<T, decltype(sizeof(T))> : std::true_type {};
+
+} // namespace detail
+
+template <class ElementType, std::size_t Extent>
+class span {
+  static_assert(std::is_object<ElementType>::value,
+                "A span's ElementType must be an object type (not a "
+                "reference type or void)");
+  static_assert(detail::is_complete<ElementType>::value,
+                "A span's ElementType must be a complete type (not a forward "
+                "declaration)");
```

Comment:
Why is this?

Reply:
See https://en.cppreference.com/w/cpp/container/span

> T 	- element type; must be a complete object type that is not an abstract class type

See also:

```c++
// detect if T is a complete type
// NB: sizeof(T) cannot be used with incomplete types
template <class, class = std::size_t>
struct is_complete : std::false_type {};

template <class T>
struct is_complete<T, decltype(sizeof(T))> : std::true_type {};
```

Reply:
In other words, this is standards specified. Ok.